File: jwt.go
package main

import (
	"context"
	"crypto/rand"
	"database/sql"
	"fmt"
	"sync"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/lib/pq"
)

type JwtClaims struct {
	UserID   int64  `json:"sub"`
	PublicID string `json:"public_id"`
	Username string `json:"username"`
	jwt.RegisteredClaims
}

type JwtService struct {
	db               *sql.DB
	secret           []byte
	issuer           string
	expiresInMinutes int
	blacklistMu      sync.RWMutex
	blacklistCache   map[string]int64
	stopCh           chan struct{}
}

func NewJwtService(db *sql.DB, secret, issuer string, expiresInMinutes int) *JwtService {
	js := &JwtService{
		db:               db,
		secret:           []byte(secret),
		issuer:           issuer,
		expiresInMinutes: expiresInMinutes,
		blacklistCache:   make(map[string]int64),
		stopCh:           make(chan struct{}),
	}
	js.loadBlacklistFromDB()
	return js
}

func (j *JwtService) loadBlacklistFromDB() {
	rows, err := j.db.Query(`SELECT jti, expires_at FROM jwt_blacklist`)
	if err != nil {
		return
	}
	defer rows.Close()

	now := time.Now().UnixMilli()
	j.blacklistMu.Lock()
	defer j.blacklistMu.Unlock()

	for rows.Next() {
		var jti string
		var expiresAt int64
		if err := rows.Scan(&jti, &expiresAt); err != nil {
			continue
		}
		if expiresAt > now {
			j.blacklistCache[jti] = expiresAt
		}
	}
}

func (j *JwtService) BlacklistToken(jti string, expiresAt int64) error {
	_, err := j.db.Exec(`
		INSERT INTO jwt_blacklist (jti, expires_at)
		VALUES ($1, $2)
		ON CONFLICT (jti) DO NOTHING`,
		jti, expiresAt,
	)
	if err == nil {
		j.blacklistMu.Lock()
		j.blacklistCache[jti] = expiresAt
		j.blacklistMu.Unlock()
	}
	return err
}

func (j *JwtService) IsBlacklisted(jti string) bool {
	j.blacklistMu.RLock()
	exp, ok := j.blacklistCache[jti]
	j.blacklistMu.RUnlock()
	return ok && exp > time.Now().UnixMilli()
}

func (j *JwtService) CleanupBlacklist() {
	now := time.Now().UnixMilli()

	// 策略：先快速扫描收集过期项（持有读锁），再批量删除（持有写锁）
	// 这样减少写锁持有时间，提高并发性能

	// 第1步：快速扫描，收集过期项（读锁，不阻塞其他读操作）
	j.blacklistMu.RLock()
	expiredItems := make([]string, 0, 100) // 预分配容量，避免频繁扩容
	for jti, exp := range j.blacklistCache {
		if exp < now {
			expiredItems = append(expiredItems, jti)
			if len(expiredItems) >= 1000 {
				// 限制单次处理数量，避免内存占用过大
				break
			}
		}
	}
	j.blacklistMu.RUnlock()

	// 第2步：批量删除（写锁，但持有时间很短）
	if len(expiredItems) > 0 {
		j.blacklistMu.Lock()
		// 双重检查：删除前再次确认是否过期
		// 防止在 RUnlock 和 Lock 之间被其他 goroutine 修改
		for _, jti := range expiredItems {
			if exp, ok := j.blacklistCache[jti]; ok && exp < now {
				delete(j.blacklistCache, jti)
			}
		}
		j.blacklistMu.Unlock()

		// 第3步：异步清理数据库（慢操作，不阻塞）
		go func(items []string) {
			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
			defer cancel()

			// 使用批量删除，提高效率
			if len(items) == 1 {
				_, _ = j.db.ExecContext(ctx, `DELETE FROM jwt_blacklist WHERE jti = $1`, items[0])
			} else {
				// PostgreSQL 支持批量删除
				_, _ = j.db.ExecContext(ctx,
					`DELETE FROM jwt_blacklist WHERE jti = ANY($1)`,
					pq.Array(items))
			}
		}(expiredItems)
	}
}

func (j *JwtService) StartCleanup(interval time.Duration) {
	go func() {
		ticker := time.NewTicker(interval)
		defer ticker.Stop()
		for {
			select {
			case <-ticker.C:
				j.CleanupBlacklist()
			case <-j.stopCh:
				return
			}
		}
	}()
}

func (j *JwtService) Stop() {
	close(j.stopCh)
}

func (j *JwtService) GenerateToken(user *User) (string, error) {
	now := time.Now()
	expiresAt := now.Add(time.Duration(j.expiresInMinutes) * time.Minute)

	jtiBytes := make([]byte, 16)
	if _, err := rand.Read(jtiBytes); err != nil {
		return "", fmt.Errorf("failed to generate token ID: %w", err)
	}
	jti := fmt.Sprintf("%x", jtiBytes)

	claims := JwtClaims{
		UserID:   user.ID,
		PublicID: user.PublicID,
		Username: user.Username,
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(expiresAt),
			IssuedAt:  jwt.NewNumericDate(now),
			NotBefore: jwt.NewNumericDate(now),
			Issuer:    j.issuer,
			Audience:  []string{j.issuer},
			ID:        jti,
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString(j.secret)
}

func (j *JwtService) ValidateToken(tokenStr string) (*JwtClaims, error) {
	token, err := jwt.ParseWithClaims(tokenStr, &JwtClaims{}, func(t *jwt.Token) (interface{}, error) {
		if _, ok := t.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", t.Header["alg"])
		}
		return j.secret, nil
	}, jwt.WithIssuer(j.issuer), jwt.WithAudience(j.issuer))
	if err != nil {
		return nil, fmt.Errorf("invalid token: %w", err)
	}

	claims, ok := token.Claims.(*JwtClaims)
	if !ok || !token.Valid {
		return nil, fmt.Errorf("invalid token claims")
	}

	return claims, nil
}

File: src/jwt.rs
use axum::extract::State;
use chrono::{Duration, Utc};
use jsonwebtoken::{decode, encode, DecodingKey, EncodingKey, Header, Validation};
use serde::{Deserialize, Serialize};
use sqlx::PgPool;
use std::collections::HashMap;
use std::sync::Arc;
use std::sync::RwLock;
use tokio::time;

const MAX_BLACKLIST_CACHE_SIZE: usize = 10000;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct JwtClaims {
    pub user_id: i64,
    pub public_id: String,
    pub username: String,
    pub exp: i64,
    pub iat: i64,
    pub nbf: i64,
    pub iss: String,
    pub aud: String,
    pub jti: String,
}

#[derive(Clone)]
pub struct JwtService {
    pool: PgPool,
    secret: Vec<u8>,
    issuer: String,
    expires_in_minutes: i64,
    blacklist_cache: Arc<RwLock<HashMap<String, i64>>>,
}

impl JwtService {
    pub fn new(pool: PgPool, secret: &str, issuer: &str, expires_in_minutes: i64) -> Self {
        let service = Self {
            pool,
            secret: secret.as_bytes().to_vec(),
            issuer: issuer.to_string(),
            expires_in_minutes,
            blacklist_cache: Arc::new(RwLock::new(HashMap::new())),
        };
        service.load_blacklist_from_db();
        service
    }

    fn load_blacklist_from_db(&self) {
        let rt = match tokio::runtime::Runtime::new() {
            Ok(rt) => rt,
            Err(e) => {
                tracing::error!("failed to create runtime for loading JWT blacklist: {}", e);
                return;
            }
        };
        
        let result = rt.block_on(async {
            let query = r#"SELECT jti, expires_at FROM jwt_blacklist"#;
            sqlx::query_as::<_, (String, i64)>(query)
                .fetch_all(&self.pool)
                .await
        });

        match result {
            Ok(rows) => {
                let now = Utc::now().timestamp_millis();
                let mut cache = HashMap::new();
                for (jti, expires_at) in rows {
                    if expires_at > now {
                        cache.insert(jti, expires_at);
                    }
                }
                self.blacklist_cache.write().unwrap().extend(cache);
            }
            Err(e) => {
                tracing::warn!("failed to load blacklist from database, cache will be empty: {}", e);
            }
        }
    }

    pub async fn blacklist_token(&self, jti: &str, expires_at: i64) -> Result<(), sqlx::Error> {
        let query = r#"
            INSERT INTO jwt_blacklist (jti, expires_at)
            VALUES ($1, $2)
            ON CONFLICT (jti) DO NOTHING"#;
        sqlx::query(query)
            .bind(jti)
            .bind(expires_at)
            .execute(&self.pool)
            .await?;

        let mut cache = self.blacklist_cache.write().unwrap();
        if cache.len() >= MAX_BLACKLIST_CACHE_SIZE {
            let now = Utc::now().timestamp_millis();
            cache.retain(|_, v| *v > now);
        }
        cache.insert(jti.to_string(), expires_at);
        Ok(())
    }

    pub fn is_blacklisted(&self, jti: &str) -> bool {
        let now = Utc::now().timestamp_millis();

        let mut cache = self.blacklist_cache.write().unwrap();
        if let Some(&exp) = cache.get(jti) {
            if exp <= now {
                cache.remove(jti);
                false
            } else {
                true
            }
        } else {
            false
        }
    }

    pub async fn cleanup_blacklist(&self) {
        let now = Utc::now().timestamp_millis();

        let expired_items: Vec<String> = {
            let cache = self.blacklist_cache.read().unwrap();
            cache
                .iter()
                .filter(|(_, exp)| **exp < now)
                .take(1000)
                .map(|(jti, _)| jti.clone())
                .collect()
        };

        if !expired_items.is_empty() {
            {
                let mut cache = self.blacklist_cache.write().unwrap();
                for jti in &expired_items {
                    if let Some(&exp) = cache.get(jti) && exp < now {
                        cache.remove(jti);
                    }
                }
            }

            let pool = self.pool.clone();
            let items = expired_items.clone();
            tokio::spawn(async move {
                if items.len() == 1 {
                    let _ = sqlx::query(r#"DELETE FROM jwt_blacklist WHERE jti = $1"#)
                        .bind(&items[0])
                        .execute(&pool)
                        .await;
                } else {
                    let _ = sqlx::query(r#"DELETE FROM jwt_blacklist WHERE jti = ANY($1)"#)
                        .bind(&items)
                        .execute(&pool)
                        .await;
                }
            });
        }
    }

    pub fn start_cleanup(&self, interval: std::time::Duration) {
        let service = self.clone();
        tokio::spawn(async move {
            let mut ticker = time::interval(interval);
            loop {
                ticker.tick().await;
                service.cleanup_blacklist().await;
            }
        });
    }

    pub fn generate_token(&self, user_id: i64, public_id: &str, username: &str) -> Result<String, String> {
        let now = Utc::now();
        let expires_at = now + Duration::minutes(self.expires_in_minutes);
        let jti = uuid::Uuid::new_v4().to_string();

        let claims = JwtClaims {
            user_id,
            public_id: public_id.to_string(),
            username: username.to_string(),
            exp: expires_at.timestamp(),
            iat: now.timestamp(),
            nbf: now.timestamp(),
            iss: self.issuer.clone(),
            aud: self.issuer.clone(),
            jti,
        };

        let token = encode(
            &Header::default(),
            &claims,
            &EncodingKey::from_secret(&self.secret),
        )
        .map_err(|e| format!("failed to generate token: {}", e))?;

        Ok(token)
    }

    pub fn validate_token(&self, token_str: &str) -> Result<JwtClaims, String> {
        let mut validation = Validation::default();
        validation.set_issuer(std::slice::from_ref(&self.issuer));
        validation.set_audience(std::slice::from_ref(&self.issuer));

        let token_data = decode::<JwtClaims>(
            token_str,
            &DecodingKey::from_secret(&self.secret),
            &validation,
        )
        .map_err(|e| format!("invalid token: {}", e))?;

        let claims = token_data.claims;

        if self.is_blacklisted(&claims.jti) {
            return Err("token is blacklisted".to_string());
        }

        Ok(claims)
    }
}

pub async fn auth_middleware(
    app_state: State<crate::AppState>,
    req: axum::http::Request<axum::body::Body>,
    next: axum::middleware::Next,
) -> axum::response::Response {
    let auth_header = req
        .headers()
        .get(axum::http::header::AUTHORIZATION)
        .and_then(|h| h.to_str().ok());

    match auth_header {
        Some(token) => {
            let token_str = token.strip_prefix("Bearer ").unwrap_or(token);

            match app_state.jwt_service.validate_token(token_str) {
                Ok(claims) => {
                    if app_state.jwt_service.is_blacklisted(&claims.jti) {
                        return axum::response::Response::builder()
                            .status(axum::http::StatusCode::UNAUTHORIZED)
                            .body(axum::body::Body::from("Token is blacklisted"))
                            .unwrap();
                    }

                    let mut req = req;
                    req.extensions_mut().insert(claims.user_id);
                    req.extensions_mut().insert(claims.public_id);
                    next.run(req).await
                }
                Err(_) => {
                    axum::response::Response::builder()
                        .status(axum::http::StatusCode::UNAUTHORIZED)
                        .body(axum::body::Body::from("Invalid token"))
                        .unwrap()
                }
            }
        }
        None => {
            axum::response::Response::builder()
                .status(axum::http::StatusCode::UNAUTHORIZED)
                .body(axum::body::Body::from("Missing token"))
                .unwrap()
        }
    }
}
